[clang-tidy] fix match for binaryOperator in ExprMutationAnalyzer for misc-const-correctness

5chmidti opened the pull request:
The `ExprMutationAnalyzer`s matcher of `binaryOperator`s
that contained the variable expr, were previously narrowing the
variable to be type dependent, when the `binaryOperator` should
have been narrowed as dependent.
The variable we are trying to find mutations for does
not need to be the dependent type, the other operand of
the `binaryOperator` could be dependent.

Fixes #57297

5chmidti committed:
[clang-tidy] fix match for binaryOperator in ExprMutationAnalyzer for misc-const-correctness

The `ExprMutationAnalyzer`s matcher of `binaryOperator`s
that contained the variable expr, were previously narrowing the
variable to be type dependent, when the `binaryOperator` should
have been narrowed as dependent.
The variable we are trying to find mutations for does
not need to be the dependent type, the other operand of
the `binaryOperator` could be dependent.

Fixes #57297

5chmidti commented:
CC @JonasToth for your work on `ExprMutationAnalyzer`

PiotrZSL commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -683,6 +683,9 @@ Bug Fixes to AST Handling
 - Fixed a bug where RecursiveASTVisitor fails to visit the
   initializer of a bitfield.
   `Issue 64916 <https://github.com/llvm/llvm-project/issues/64916>`_
+- Fixed a bug where ``ExprMutationAnalyzer`` did not find a potential mutation
```
I think this should be under static analyzer changes or not included at all.

5chmidti replied:
Done

5chmidti committed:
rm release note comment from clang specific docs

5chmidti commented:
The top-level Linux and Windows tests work, but inside the `Trigger Build`/`clang-ci` run, the windows test step fails. I'm not sure what the difference is between these two Windows tests.

jcsxky commented:
> The top-level Linux and Windows tests work, but inside the `Trigger Build`/`clang-ci` run, the windows test step fails. I'm not sure what the difference is between these two Windows tests.

This is interesting! Generally speaking, AST of the code in Windows and Linux platform is identical with clang compiled. Probably need debuging in Windows.

5chmidti commented:
I found the issue, but I have not looked for a solution.
See https://godbolt.org/z/or4EnMfWj and #71203.
When using clang with `-target` to specify `x86_64-pc-win32` and `x86_64-linux`, the body for the function `f` is `NULL` in the AST with the windows target. This only happens for `-target x86_64-windows` pre `c++20`.

Although the tested code does not directly depend on being `c++20` (it is valid in `c++11`), maybe I constrain the tests in this pr to be in `c++20` mode until the issue is fixed? What do you think?

jcsxky commented:
how about adding a function call like `f<int>()`? will this can prevent clang to emit empty function body?

5chmidti committed:
fix windows CI failure

- explicitly instantiate template
- provide operator int() to Stream to ensure << is a BinaryOperator,
  not a CXXOperatorCallExpr

5chmidti commented:
> how about adding a function call like `f<int>()`? will this can prevent clang to emit empty function body?

Yep, thanks. I went this route, but I had to define a conversion operator to compile and keep the `<<` a `BinaryOperator`. The CI now passes

PiotrZSL commented:
In tests we usually use `-fno-delayed-template-parsing` to overcome this issue with templates and windows.

5chmidti committed:
test with `-fno-delayed-template-parsing` instead

5chmidti commented:
> In tests we usually use `-fno-delayed-template-parsing` to overcome this issue with templates and windows.

Done. I changed the test in `ExprMutationAnalyzerTest.cpp` to use the flag, `const-correctness-templates.cpp` already had it, and I changed the tested code back to what it was before (-formatting).

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -308,6 +308,11 @@ Changes in existing checks
   <clang-tidy/checks/misc/const-correctness>` check to avoid false positive when
   using pointer to member function.
 
+- Improved :doc:`misc-const-correctness
```
there is already entry for this check, just merge changes.

5chmidti replied:
Done. I have also reworded the release note, I think it reads better now.

5chmidti committed:
combine release note entries

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -307,11 +307,9 @@ Changes in existing checks
-  <clang-tidy/checks/misc/const-correctness>` check to not warn on uses in
-  type-dependent binary operators, when the variable that is being
-  looked at, is not the dependent operand.
+  Additionally, the check no longer emits a diagnostic when
```
Looks fine, but wrap this after "to member function."

5chmidti replied:
Like this?

5chmidti committed:
don't start a new line for a sentence
